pydantic_aioredis/utils.py

```diff
@@ -1,6 +1,308 @@
+        self.revert()
+
+    def apply(self, loop=None):
+        """Patch asyncio to make its event loop reentrant."""
+        if not self.patched:
+            self.patch_asyncio()
+            self.patch_policy()
+            self.patch_tornado()
+
+            loop = loop or asyncio.get_event_loop()
+            self.patch_loop(loop)
+            self.patched = True
+
+    def revert(self):
+        if self.patched:
+            for loop in self.orig_loop_attrs:
+                self.unpatch_loop(loop)
+            self.unpatch_tornado()
+            self.unpatch_policy()
+            self.unpatch_asyncio()
+            self.patched = False
+
+    def patch_asyncio(self):
+        """Patch asyncio module to use pure Python tasks and futures."""
+
+        def run(main, *, debug=False):
+            loop = asyncio.get_event_loop()
+            loop.set_debug(debug)
+            task = asyncio.ensure_future(main)
+            try:
+                return loop.run_until_complete(task)
+            finally:
+                if not task.done():
+                    task.cancel()
+                    with suppress(asyncio.CancelledError):
+                        loop.run_until_complete(task)
+
+        def _get_event_loop(stacklevel=3):
+            return events._get_running_loop() or events.get_event_loop_policy().get_event_loop()
+
+        # Use module level _current_tasks, all_tasks and patch run method.
+        if getattr(asyncio, "_nest_patched", False):
+            return
+        self.orig_tasks = [asyncio.Task, asyncio.tasks._CTask, asyncio.tasks.Task]
+        asyncio.Task = asyncio.tasks._CTask = asyncio.tasks.Task = asyncio.tasks._PyTask
+        self.orig_futures = [asyncio.Future, asyncio.futures._CFuture, asyncio.futures.Future]
+        asyncio.Future = asyncio.futures._CFuture = asyncio.futures.Future = asyncio.futures._PyFuture
+        self.orig_get_loops = {
+            "events_get_event_loop": events.get_event_loop,
+            "asyncio_get_event_loop": asyncio.get_event_loop,
+        }
+        if sys.version_info >= (3, 10, 0) and sys.version_info < (3, 12, 0):
+            self.orig_get_loops["events__get_event_loop"] = events._get_event_loop
+        events._get_event_loop = events.get_event_loop = asyncio.get_event_loop = _get_event_loop
+        self.orig_run = asyncio.run
+        asyncio.run = run
+        asyncio._nest_patched = True
+
+    def unpatch_asyncio(self):
+        if self.orig_run:
+            asyncio.run = self.orig_run
+            asyncio._nest_patched = False
+            (asyncio.Task, asyncio.tasks._CTask, asyncio.tasks.Task) = self.orig_tasks
+            (asyncio.Future, asyncio.futures._CFuture, asyncio.futures.Future) = self.orig_futures
+            if sys.version_info >= (3, 9, 0):
+                for key, value in self.orig_get_loops.items():
+                    setattr(asyncio if key.startswith("asyncio") else events, key.split("_")[-1], value)
```

Comment:
@andrewthetechie @CharlieJiangXXX just in case, should this be `key.split("_", 1)`?

```suggestion
                    setattr(asyncio if key.startswith("asyncio") else events, key.split("_", 1)[-1], value)
```

Reply:
It is unfortunate that I cannot put comment to https://github.com/erdewit/nest_asyncio/pull/88 anymore